### PR TITLE
Macros

### PR DESCRIPTION
**Pull request by Splagoon**

### CODE REVIEW

File: lib/asm510/generator.ex [hunk 1]
```diff
@@ -135,22 +165,225 @@ defmodule ASM510.Generator do
...
+        # defined during the course of the macro
+        |> Map.put("\\#{arg_name}", fn line, env ->
+          case Map.fetch(env, name) do
+            :error ->
+              {:error, line, {:undefined_symbol, name}}
+
+            result ->
+              result
+          end
+        end)
```
Comment (SupernaviX) [inline lib/asm510/generator.ex:243]: Does this lazy evaluation allow you to reference labels defined after the macro?

File: test/input/test.s
Comment (SupernaviX) [inline test/input/test.s:1]: nice! abstraction

**Review by SupernaviX — COMMENTED** [review]
I'm not allowed to approve pull requests

File: lib/asm510/generator.ex (hunk 1 above)
Comment (Splagoon) [reply lib/asm510/generator.ex]: This lazy evaluation enables the following:
```
.macro define x
.set 'x, 1
.set 'x, \x + 1
.endm

define foo
```

Although `foo` is undefined when the macro is called, the macro doesn't refer to its value until after it's defined, at which point the lazy evaluation will succeed. This doesn't help if its value is defined later in the program though; only if it's defined between the start of the macro and when the macro gets its value.

File: README.md [hunk 2]
```diff
@@ -85,12 +183,15 @@ Sets the location counter to the value of _location_. The location counter can b
...
 
 For example, assembling
+
 ```asm
 .rept 3
 .word 1
 .endr
 ```
+
 is equiavlent to assembling
```
Comment (SupernaviX) [inline README.md:193]: *equivalent

**Pull request merged**